File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/loader/http/TransportTask.java
package com.example.alexeyglushkov.taskmanager.loader.http;

import androidx.annotation.NonNull;

import com.example.alexeyglushkov.streamlib.progress.ProgressUpdater;
import com.example.alexeyglushkov.taskmanager.loader.transport.TaskTransport;
import com.example.alexeyglushkov.taskmanager.task.SimpleTask;

public class TransportTask extends SimpleTask implements TaskTransport.Listener {
    private TaskTransport transport;

    public TransportTask() {
        super();
    }

    public TransportTask(TaskTransport transport) {
        super();
        setTransport(transport);
    }

    public void setTransport(@NonNull TaskTransport transport) {
        TaskTransport oldTransport = getTransport();
        if (oldTransport != null && oldTransport.getListener() == this) {
            oldTransport.setListener(null);
        }

        this.transport = transport;
        transport.setListener(this);

        String transportId = transport.getId();
        if (transportId != null) {
            setTaskId(transportId);
        }
    }

    public TaskTransport getTransport() {
        return transport;
    }

    public void startTask(Callback callback) {
        super.startTask(callback);

        transport.start();

        if (needCancelTask) {
            setIsCancelled();

        } else {
            Error e = transport.getError();
            Object d = transport.getData();

            if (e != null) {
                setError(e);
            } else {
                setTaskResult(d);
            }
        }

        getPrivate().handleTaskCompletion(callback);
    }

    public void setError(Error error) {
        getPrivate().setTaskError(error);
    }

    @Override
    public boolean canBeCancelledImmediately() {
        return true;
    }

    @Override
    public void clear() {
        super.clear();

        transport.setListener(null);
        transport = null;
    }

    @Override
    public void cancelTask(Object info) {
        ProgressUpdater progressUpdater = transport.getProgressUpdater();

        if (progressUpdater != null) {
            transport.cancel();
            progressUpdater.cancel(info); // cancel will call this method again
        } else {
            super.cancelTask(info);
        }
    }

    // TaskTransport.Listener

    @Override
    public ProgressUpdater getProgressUpdater(TaskTransport transport, float size) {
        return createProgressUpdater(size);
    }

    @Override
    public boolean needCancel(TaskTransport transport) {
        return getNeedCancelTask();
    }
}

File: taskmanager_httptask/src/main/java/com/aglushkov/taskmanager_http/loader/http/TransportTask.kt
package com.aglushkov.taskmanager_http.loader.http

import com.example.alexeyglushkov.streamlib.progress.ProgressUpdater
import com.aglushkov.taskmanager_http.loader.transport.TaskTransport
import com.example.alexeyglushkov.taskmanager.task.SimpleTask

class TransportTask : SimpleTask, TaskTransport.Listener {
    private var _transport: TaskTransport? = null
    var transport: TaskTransport
        get() {
           return _transport!!
        }
        set(value) {
            if (_transport != null && _transport?.listener === this) {
                transport.listener = null
            }

            _transport = value
            transport.listener = this

            val transportId = transport.id
            if (transportId != null) {
                taskId = transportId
            }
        }

    constructor() : super()

    constructor(transport: TaskTransport) : super() {
        this.transport = transport
    }

    override suspend fun startTask() {
        transport.start()

        if (needCancelTask) {
            setIsCancelled()

        } else {
            val e = transport.error
            val d = transport.data

            if (e != null) {
                setError(e)
            } else {
                taskResult = d
            }
        }

        private.handleTaskCompletion()
    }

    fun setError(error: Error) {
        private.taskError = error
    }

    override fun canBeCancelledImmediately(): Boolean {
        return true
    }

    override fun clear() {
        super.clear()
        transport.clear()
    }

    override fun cancelTask(info: Any?) {
        val progressUpdater = transport.progressUpdater

        if (progressUpdater != null) {
            transport.cancel()
            progressUpdater.cancel(info) // cancel will call this method again
        } else {
            super.cancelTask(info)
        }
    }

    // TaskTransport.Listener

    override fun getProgressUpdater(transport: TaskTransport, size: Float): ProgressUpdater {
        return createProgressUpdater(size)
    }

    override fun needCancel(transport: TaskTransport): Boolean {
        return needCancelTask
    }
}
